### PR TITLE
More specific TemplateStringsArray type for tagged templates

### LINKED ISSUE #31422: const TemplateStringsArray for TaggedTemplateExpression
## Search Terms

TemplateStringsArray, TaggedTemplateExpression

## Suggestion

There shouldn't be a type `TemplateStringsArray`, it should be a `const string tuple type`, so we can write code:

```ts
interface SQL<TSA, VS> {
  texts: TSA;
  values: VS;
}

function sql<TSA extends readonly string[], VS extends any[]>(texts:TSA, ...values: VS): SQL<TSA, VS> {
  return { texts, values };
}

// then
let s: SQL<['select * from person where a=', ' and b=', ''], [number, Date]> = sql`select * from person where a=${1} and b=${new Date()}`;
```

## Use Cases

Just like in the above code, we can test the sql at compile time or use a TypeScript Language Service Plugin *(in fact, I'm writing it[ts-sql-plugin](https://github.com/xialvjun/ts-sql-plugin) and then come across this problem)*.

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [ ] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

There is another issue is alike. https://github.com/microsoft/TypeScript/issues/16552#issuecomment-492919476

### LINKED ISSUE #33304: Type safety with `TemplateStringsArray` and tag functions
## Search Terms

`TemplateStringsArray`, `type`, `safety`, `generic`, `tag`, `function`

## Suggestion

Hello, I'd like to know how could I achieve type safety with `TemplateStringsArray` and tag functions or make such scenarios possible.

## Use Cases

I would use `TemplateStringsArray` together with the 

```ts
tagFunction`<type_safe_accessor>`
```

Currently, there's no way to do this, since

1. the definition of `TemplateStringsArray` is not generic

```ts
// source code:
// TypeScript/src/lib/es5.d.ts
// TypeScript/lib/lib.es5.d.ts

// local install:
// /usr/lib/code/extensions/node_modules/typescript/lib/lib.es5.d.ts

interface TemplateStringsArray extends ReadonlyArray<string> {
    readonly raw: ReadonlyArray<string>;
}
```

AND 

2. the usage of

```ts
tagFunction`<type_safe_accessor>`
```

gives the following error, completely preventing the type-safe usage:

```ts
Argument of type '{}' is not assignable to parameter of type keyof TypeAndValueObject | TemplateStringsArray<keyof TypeAndValueObject>'.
  Type '{}' is missing the following properties from type 'TemplateStringsArray<keyof TypeAndValueObject>': raw, length, concat, join, and 19 more. ts(2345)
```

### The working case

I have some type-safe `i18n` translations:

```ts
// Dictionary.ts
export interface Dictionary {
	"Hello": string;
	"Click to see more": string;
	"Today is": (day: string) => string;
}

// en.ts
import { Dictionary } from "./Dictionary";

export const en: Dictionary = {
	"Hello": "Hello",
	"Click to see more": "Click to see more",
	"Today is": (day: string) => `Today is ${day}`,
};

// lt.ts
import { Dictionary } from "./Dictionary";

export const lt: Dictionary = {
	"Hello": "Sveiki",
	"Click to see more": "Paspauskite, kad pamatytumėte daugiau",
	"Today is": (day: string) => `Šiandien yra ${day}`,
};
```

```ts
// i18n.ts
import { Dictionary } from "./Dictionary";
import { en } from "./en";
import { lt } from "./lt";

export interface ITranslations {
	en: Dictionary;
	lt: Dictionary;
}

export const translations: ITranslations = {
	en: en,
	lt: lt,
};

// "en" | "lt"
export type ILang = keyof ITranslations;
```

I have a function to get the translations:

```ts
import { ILang, translations } from "./i18n";
import { Dictionary } from "./Dictionary";

const currentLang: ILang = "en";

const dictionary: Dictionary = translations[currentLang];

export const selectTranslation = <K extends keyof Dictionary>(key: K): Dictionary[K] => {
	const translationText: Dictionary[K] = dictionary[key];
	return translationText;
};
```

And I can safely use it with type-safe translations in the following fashion:

```ts
// someFile.ts
import { selectTranslation } from "../selectTranslation.ts";

/**
 * Type-checks, auto-completes etc. the strings from the `Dictionary`
 */
const someText: string = selectTranslation("Hello");
```

### The problem

However, now I'd like to use the template string (template literal) syntax, paired with the tag function to achieve the same functionality, so I improve the `selectTranslation` function:

```ts
// selectTranslation.ts
import { ILang, translations } from "./i18n";
import { Dictionary } from "./Dictionary";

const currentLang: ILang = "en";

const dictionary: Dictionary = translations[currentLang];

export const selectTranslation = <K extends keyof Dictionary>(key: K | TemplateStringArray): Dictionary[K] => {
	let realKey: K;

	if (Array.isArray(key)) {
		realKey = key[0];
	} else {
		realKey = key; // error:
		/** 
		 * Type 'K | TemplateStringsArray<string>' is not assignable to type 'K'.
  		 * Type 'TemplateStringsArray<string>' is not assignable to type 'K'. ts(2322)
		*/
	}

	const translationText: Dictionary[K] = dictionary[realKey];
	return translationText;
};
```

```ts
// anotherFile.ts
import { selectTranslation } from "../selectTranslation.ts";

/**
 * Does NOT type-check and gives the previously mentioned error:
 * 
 * Argument of type '{}' is not assignable to parameter of type K | TemplateStringsArray<K>'.
 * Type '{}' is missing the following properties from type 'TemplateStringsArray<K>': raw, length, concat, join, and 19 more. ts(2345)
 */
const someText: string = selectTranslation`Hello`; // error
```

## Possible solutions

I've tried 3 different solutions, neither of them giving me the results that I want:

### a) Change the `TemplateStringsArray` to be a generic like so

```ts
// source code:
// TypeScript/src/lib/es5.d.ts
// TypeScript/lib/lib.es5.d.ts

// local install:
// /usr/lib/code/extensions/node_modules/typescript/lib/lib.es5.d.ts

interface TemplateStringsArray<T = string> extends ReadonlyArray<T> {
    readonly raw: ReadonlyArray<T>;
}
```

and used it like so

```diff
// selectTranslation.ts
-export const selectTranslation = <K extends keyof Dictionary>(key: K | TemplateStringsArray): Dictionary[K] => {
+export const selectTranslation = <K extends keyof Dictionary>(key: K | TemplateStringsArray<K>): Dictionary[K] => {
```

however, the same problems persisted - the casting from `key` to `realKey` failed
AND the tagged function usage still failed

```ts
// selectTranslation.ts

if (Array.isArray(key)) {
	realKey = key[0];
} else {
	realKey = key; // still errors
}

// anotherFile.ts
const someText: string = selectTranslation`Hello`; // still errors
```

### b) Instead of using `TemplateStringsArray`, just use `Array<K>`

```diff
// selectTranslation.ts
-export const selectTranslation = <K extends keyof Dictionary>(key: K | TemplateStringsArray): Dictionary[K] => {
+export const selectTranslation = <K extends keyof Dictionary>(key: K | Array<K>): Dictionary[K] => {
```

the first problem of casting from `key` to `realKey` disappeared,
BUT the second one still remained

```ts
// selectTranslation.ts

if (Array.isArray(key)) {
	realKey = key[0];
} else {
	realKey = key; // all cool now
}

// anotherFile.ts
const someText: string = selectTranslation`Hello`; // still errors!!!
```

### c) Just use `any`

```diff
// selectTranslation.ts
-export const selectTranslation = <K extends keyof Dictionary>(key: K | TemplateStringsArray): Dictionary[K] => {
+export const selectTranslation = <K extends keyof Dictionary>(key: K | any): Dictionary[K] => {
```

which then allows me to use the tag function, **BUT** there's NO type-safety, autocompletions etc., making it practically useless.

## TL;DR:

Neither of the solutions helped - I'm still unable to use the `selectTranslation` function as a tag function with type safety.

Is there any way to make it possible?

Reminder - we have 2 problems here:

1. (less important since it can be avoided by using `Array<K>`) `TemplateStringsArray` does not work like it should (or I'm using it wrong), even when used as a generic 
2. (more important) tag functions cannot have type-safe parameters? (PLEASE fix)

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

---

I'm happy to help if you have any further questions.

### PR DESCRIPTION
**Pull request by rbuckton**
This introduces a more specific type over `TemplateStingsArray` when invoking a tagged template. This would then allow developers to interpolate the literal types provided to the tagged template invocation, along with the provided arguments:

```ts

type Interpolate<T extends readonly string[], A extends any[], R extends string = ''> =
    T extends readonly [infer TH extends string, ...infer TT extends readonly string[]] ?
        A extends [infer AH extends string, ...infer AT extends string[]] ?
            Interpolate<TT, AT, `${R}${TH}${AH}`> :
            Interpolate<TT, [], `${R}${TH}`> :
        R;

declare function interp<
  T extends TemplateStringsArray,
  A extends string[]
>(strs: T, ...args: A): Interpolate<T, A>;

const a = interp`a${"b"}c`; // "abc"
const b = interp`1${2}3`; // "123"
```

The type we manufacture for the tagged template invocation is an instantiation of `TemplateStringsArrayOf<Cooked, Raw>`:

```ts
type TemplateStringsArrayOf<
  Cooked extends readonly string[],
  Raw extends readonly string[]
> =
  Cooked & { readonly raw: Raw };
```

This would then allow you to perform custom interpolation against either the "cooked" values (where escapes are normalized), and "raw" values (where escapes are preserved):

```ts
declare function raw<
  T extends TemplateStringsArray,
  A extends string[]
>(strs: T, ...args: A): Interpolate<T["raw"], A>;

const a = raw`a\n${"b"}\nc`; // "a\\nb\\nc"
```

In theory, more complex interpolation could also be achieved, but that is an exercise left to the reader:

```ts
type QueryResult<T extends readonly string[], A extends any[]> =
   // some type that parses T and interpolates A...
  todo;

declare function query<T extends TemplateStringsArray, A extends any[]>(strs: T, ...args: A): QueryResult<T, A>;

interface Person { firstName: string, lastName: string }
declare const people: Iterable<Person>;

const result = query`
  for x in ${people}
  where x.firstName === ${"Bob"}
  select x.lastName
`;

result; // Iterable<string>
```

To support interpolation over a `TemplateStringsArray`, I also needed to make a change to `inferFromObjectTypes`. Prior to this change you could not infer to tuple element positions if the source type was an *intersection* of a tuple and another type:

```ts
type X<T extends any[]> = T extends [infer H, ...infer _] ? H : never;

// before
type T1 = X<[1, 2]>; // 1 (ok)
type T2 = X<[1, 2] & { foo: true }>; // never (😞)
```

However, this becomes a problem when trying to pick apart a generic `TemplateStringsArrayOf` type. To address this, this PR makes a small change in how we perform this inference. Nothing changes when a source type is already an array or tuple type, but if the source type is instead an *intersection* of an array or tuple type and another type, we will also now permit tuple element inference as long as the intersected types do not shadow tuple-specific properties (i.e., numeric indexers, numeric-stringed properties, or the `"length"` property):

```ts
type X<T extends any[]> = T extends [infer H, ...infer _] ? H : never;

// after
type T1 = X<[1, 2]>; // 1 (ok)
type T2 = X<[1, 2] & { foo: true }>; // 1 (ok)
```

Fixes #33304
Fixes #31422
Related #29432
Related #16551

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -21171,6 +21177,53 @@ namespace ts {
...
+
+            // pick an intersection constituent if it is an array or tuple type, but only
+            // if any non-array, non-tuple constituents that follow it do not shadow any
+            // tuple-specific members.
+            let arrayOrTuple: TypeReference | undefined;
+            let constituents: Type[] | undefined;
+            for (const constituent of (type as IntersectionType).types) {
+                if (isArrayOrTupleType(constituent)) {
+                    arrayOrTuple = constituent;
+                    constituents = undefined;
```
Comment (weswigham) [inline src/compiler/checker.ts:21197]: ? Isn't this kinda oddly order dependent? `{a} & []` eliminates `{a}` from consideration as a constituent that may have a numeric property, but `[] & {a}` does not.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -21171,6 +21177,53 @@ namespace ts {
...
+            }
+
+            // pick an intersection constituent if it is an array or tuple type, but only
+            // if any non-array, non-tuple constituents that follow it do not shadow any
+            // tuple-specific members.
+            let arrayOrTuple: TypeReference | undefined;
+            let constituents: Type[] | undefined;
+            for (const constituent of (type as IntersectionType).types) {
+                if (isArrayOrTupleType(constituent)) {
+                    arrayOrTuple = constituent;
```
Comment (weswigham) [inline src/compiler/checker.ts:21196]: Also this always just picks the last array or tuple type, making it order-dependent which tuple is picked, which seems bad. We should either have tuple-intersecting behavior or return `undefined` when there are multiple arrays/tuples, IMO.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -35682,6 +35766,9 @@ namespace ts {
         }
 
         function checkTypeReferenceNode(node: TypeReferenceNode | ExpressionWithTypeArguments) {
+            if (isTypeReferenceNode(node) && isIdentifier(node.typeName) && node.typeName.escapedText === "Scenario") {
+                debugger;
```
Comment (weswigham) [inline src/compiler/checker.ts:35770]: `debugger;`

**Review by weswigham — CHANGES_REQUESTED** [review]
I like most of this change, it's not too big, either; but there's some intersection-order-dependence in the inference code I think we should remove, and a `debugger` statement that obviously needs to go. 😄

File: src/compiler/checker.ts (hunk 1 above)
Comment (rbuckton) [reply src/compiler/checker.ts:21197]: We are already order dependent for this case: [Playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8UBOECGATA9gOwDYigG0ADKAEgG8BnYJAS2wHMBfYgXQG4AoUSKAIQTI0WPARKUa9JsyjtuXdBADGuVCijKcNKKgBGALlhQAZAO5LV66Fuw69qI4LMwFAejdRsmYFExIlJCglSGwlbF8ACwgUADp4rlsdAA8ARiF9QgAGTigoDyg0o1JJWgYWUjNiUukWOUTtX2SAJiEHbNz8zxbi8moymUq5GvLZYi4uAv9A4IhQ8KiYiHjYnnBoAA0AHgAVKAhk4AgwqmEMHHwoKXLCNgA+IT2Do5Oz0UvCBgAzGKgACQANFAVt9fgB9NhQAD8-ygRmwEAAbjFuLxoHtENs4GZ+HcOF0oDtiiMZPUgA)

File: src/compiler/checker.ts (hunk 2 above)
Comment (rbuckton) [reply src/compiler/checker.ts:21196]: We are already order dependent for this case: [Playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8UBOECGATA9gOwDYigG0ADKAEgG8BnYJAS2wHMBfYgXQG4AoUSKAIQTI0WPARKUa9JsyjtuXdBADGuVCijKcNKKgBGALlhQAZAO5LV66Fuw69qI4LMwFAejdRsmYFExIlJCglSGwlbF8ACwgUADp4rlsdAA8ARiF9QgAGTigoDyg0o1JJWgYWUjNiUukWOUTtX2SAJiEHbNz8zxbi8moymUq5GvLZYi4uAv9A4IhQ8KiYiHjYnnBoAA0AHgAVKAhk4AgwqmEMHHwoKXLCNgA+IT2Do5Oz0UvCBgAzGKgACQANFAVt9fgB9NhQAD8-ygRmwEAAbjFuLxoHtENs4GZ+HcOF0oDtiiMZPUgA)

File: src/compiler/checker.ts (hunk 1 above)
Comment (rbuckton) [reply src/compiler/checker.ts:21197]: This has changed in the latest update.

File: src/compiler/checker.ts (hunk 2 above)
Comment (rbuckton) [reply src/compiler/checker.ts:21196]: I don't think intersecting the tuples is viable here, as it introduces far too much complexity. I'm leaning towards what appears to be the current behavior in this case (i.e., picking the right-most), though another viable alternative would be to pick neither if there are more than one (since they would overlap).

**Comment by rbuckton** [comment]
I tinkered with an example of a tail-recursive parser for `jest.each` [here](https://gist.github.com/rbuckton/cee71acbe68a3fac8b0b4babae58213a), though it can probably be drastically simplified.

File: src/compiler/checker.ts (hunk 2 above)
Comment (weswigham) [reply src/compiler/checker.ts:21196]: Picking neither if there's more than one is often what we do in other situations, and order-dependence in inference is bad anywhere it crops up, since it means inference results can shift with file ordering, which is a pain in the butt bug to track down when it's reported later. Our existing left-preferring behavior among multiple matching inferences is already a big "this really should be fixed"; I'd much prefer to bring the cases that introduce order dependence down if possible.

**Comment by rbuckton** [comment]
Here's a much simpler version of Jest's `each`:

```ts
type IDStart =
    | "a" | "b" | "c" | "d" | "e" | "f" | "g" | "h" | "i" | "j" | "k" | "l" | "m" | "n" | "o" | "p" | "q" | "r" | "s" | "t" | "u" | "v" | "w" | "x" | "y" | "z"
    | "A" | "B" | "C" | "D" | "E" | "F" | "G" | "H" | "I" | "J" | "K" | "L" | "M" | "N" | "O" | "P" | "Q" | "R" | "S" | "T" | "U" | "V" | "W" | "X" | "Y" | "Z"
    | "$" | "_"
    ;

type IDPart =
    | "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9"
    | IDStart
    ;

type Trim<S extends string, Chars extends string = " " | "\t" | "\v" | "\n"> =
    S extends (Chars extends "" ? S : never) ? S :
    S extends `${Chars}${infer R}` ? Trim<R, Chars> :
    S extends `${infer R}${Chars}` ? Trim<R, Chars> :
    S;

type ParseHeaders<S extends string> =
    S extends "|" ? [] :
    S extends `${infer H}|${infer T}` ? [Trim<H>, ...ParseHeaders<T>] :
    [Trim<S>];

type CheckHeader<TIn extends string, TWork extends string = TIn, Valid extends IDPart = IDStart> =
    TWork extends `${infer Ch}${infer R}` ?
        Ch extends Valid ? CheckHeader<TIn, R, IDPart> :
            JestEachParseError<`Invalid identifier: '${TIn}'`> :
    TIn extends '' ? JestEachParseError<`Identifier expected.`> :
        TIn;

type CheckHeaders<TIn extends string[], TWork extends string[] = TIn> =
    TWork extends [infer TH extends string, ...infer TT extends string[]] ?
        CheckHeader<TH> extends infer E extends JestEachParseError<any> ? E :
            CheckHeaders<TIn, TT> :
        TIn;

type ParseRows<A extends any[], S extends readonly string[], Row extends any[] = [], Rows extends any[][] = []> =
    [A, S] extends [[infer AH, ...infer AT], readonly [infer TH extends string, ...infer TT extends string[]]] ?
        Trim<TH, " " | "\t" | "\v"> extends "|" ? ParseRows<AT, TT, [...Row, AH], Rows> :
        Trim<TH, " " | "\t" | "\v"> extends "\n" | "" ? ParseRows<AT, TT, [], [...Rows, [...Row, AH]]> :
            JestEachParseError<`Unexpected character: '${Trim<TH>}'`> :
    [A, S] extends [[], readonly []] ? Rows :
        JestEachParseError<`Mismatched elements`>;

type JestEachArgument<Headers extends string[] | JestEachParseError<any>, Rows extends any[][] | JestEachParseError<any>> =
    Headers extends string[] ?
        Rows extends any[][] ?
            {
                [P1 in keyof Rows]: {
                    [P2 in keyof Headers as P2 extends `${number}` ? Headers[P2] : never]:
                        P2 extends keyof Rows[P1] ? Rows[P1][P2] : undefined;
                };
            }[number] :
            Rows :
        Headers;

type JestEachFunction<Arg> =
    Arg extends JestEachParseError<any> ? Arg :
    (name: string, cb: (arg: Arg) => void, timeout?: number) => void;

declare const JestEachParseError: unique symbol;

interface JestEachParseError<Message extends string> {
    [JestEachParseError]: Message;
}

type JestEach<T extends readonly string[], A extends any[]> =
    T extends readonly [infer TH extends string, ...infer TT extends readonly string[]] ?
        JestEachFunction<JestEachArgument<CheckHeaders<ParseHeaders<TH>>, ParseRows<A, TT>>> :
        JestEachParseError<`Mismatched elements`>;

declare function each<T extends readonly string[], A extends (string | symbol | number | bigint | boolean | null | undefined | object)[]>(strs: T, ...args: A): JestEach<T, A>;

// jest
it.each`
    foo    | bar
    ${"a"} | ${"b"}
    ${"a"} | ${"c"}
`("test", ({ foo, bar }) => {
//           ^? any
//                ^? any
});

each`
    foo    | bar
    ${"a"} | ${"b"}
    ${"c"} | ${"d"}
`("test", ({ foo, bar }) => {
//           ^? "a" | "c"
//                ^? "b" | "d"
});
```

**Review by weswigham — DISMISSED** [review]
I'd prefer the order dependence removed from inference, and I'd love to see that `.each` example in the test baselines (it probably exercises a bunch that we have minimal coverage of in concert, and would be useful to know we don't regress it), but I'm willing to approve this now, even if I'd like to see a bit more.

**Comment by rbuckton** [comment]
There is no longer any order dependence. The behavior for inferring from an intersection of two tuple types now falls back to pre-existing behavior.

**Comment by rbuckton** [comment]
Also, I've added a test for Jest's `each` based on the example above.

**Comment by rbuckton** [comment]
@DanielRosenwasser: Are we comfortable taking this for 4.8 or would you rather wait?

**Comment by rbuckton** [comment]
@typescript-bot perf test
@typescript-bot run dt
@typescript-bot test this
@typescript-bot user test this

**Comment by rbuckton** [comment]
@typescript-bot perf test

**Comment by trusktr** [comment]
[reference playground](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220627#code/C4TwDgpgBAkgdsCAnMB7ANgQ0QHgCpQQAeicAJgM5RISZmpzohQXBICWcA5gNoC6AGigBBQiQjkqmOCH5CASmNKUWbTlygBeKAHIdAPi0BYAFBRzUAsWVUadBkyg9OAM2SWAEkokrWHbkIAdMGu7nhW4pLUtPSMzH7q-HxQAPymFhki3lHOcG5IIl7WPlQJAVDBgaEFwhE2qv68fMlpZpnt8IgoGNgQ+HhCtUIABgAkAN7yAL4TeB4z48Lzw4YAXOntmZ3IaFi44UJyUGOTC3NTK1Drbe3yANympmQQAMZYNFAuAK5wL8DsDCgnC6YBwGzqJUsEAAtmA9hAAMpqbgUYRIJCYEACDaiYpRMpNUz6AAUfgoq0sQWCmCQXHJIgAlBTtt14fhBvoHiZTC8GKwoJgtECEDthpgJgAiABGEqmL2GdygAHolVAJZgpS8JTy+cAoFKhcDRQBGCYAJimAGYFcrVRLjWbLRKgA)

**Comment by MartinJohns** [comment]
This is a feature I've been waiting for a long time, but I forgot why. 😄

**Comment by harrysolovay** [comment]
Hi, this PR looks fantastic (I've been hoping for this feature for a long time)! Wondering what the status is?

**Comment by SieR-VR** [comment]
Is there any progress for this feature?

**Comment by mzinner** [comment]
I hope the pending reviews can be done soon so we will see this released soon. This would open the door to a lot of interesting things.

**Comment by rbuckton** [comment]
This PR is currently on hold. The main use cases for this feature require a fair amount of complexity that is likely to run any implementation against the recursion depth limiter. Such types are expensive to compute and check, and often result in degraded performance in an editor. It's unlikely that we will ship a feature that introduces this much complexity without first making significant improvements in overall type checking performance.

I do plan to keep this PR alive and will periodically update it, but I do not foresee it shipping in the near future.

**Comment by NWYLZW** [comment]
Another year has passed, and version 5.x has now been released. This version has optimized the performance of comparison. Is it possible to consider introducing this feature in the new version? This helps us a lot, thank you very much for your work.

**Comment by easrng** [comment]
i wrote a [working type level parser](https://github.com/easrng/elements/blob/main/src/_types.ts) for my library's html template string DSL so it can be typesafe but it's useless without this PR. JSX isn't an option due to lack of ability to cache static parts with default tooling. i'd love to see this merged.

**Comment by aradalvand** [comment]
@rbuckton Any updates? Aren't v5's perf enhancements sufficient now?

**Comment by rbuckton** [comment]
> @rbuckton Any updates? Aren't v5's perf enhancements sufficient now?

No, unfortunately. The performance concerns we have are related to the complex conditional types and inference that are necessary to "parse" a string literal type, and that still isn't very efficient at the moment.

**Comment by ssalbdivad** [comment]
@rbuckton Has [ArkType](https://github.com/arktypeio/arktype) not proven it can be done efficiently at this point?

Even with the overhead of type-level string parsing (including syntactic and semantic validation), types are often 3-10x less expensive than equivalent Zod.

It would be a shame not to ship what would be an awesome feature just because it is possible to use inefficiently.

**Comment by porsager** [comment]
We need this merged yesterday ❤️ Typescript users are harassing me into changing my library APIs to be different just because they can't do what this PR would allow them. 

Why not let the resposibility of making fast implementations, and avoiding slow ones be up to the users?

**Comment by ariofrio** [comment]
> This PR is currently on hold. The main use cases for this feature require a fair amount of complexity that is likely to run any implementation against the recursion depth limiter. Such types are expensive to compute and check, and often result in degraded performance in an editor. It's unlikely that we will ship a feature that introduces this much complexity without first making significant improvements in overall type checking performance.
> 
> I do plan to keep this PR alive and will periodically update it, but I do not foresee it shipping in the near future.

> > @rbuckton Any updates? Aren't v5's perf enhancements sufficient now?
> 
> No, unfortunately. The performance concerns we have are related to the complex conditional types and inference that are necessary to "parse" a string literal type, and that still isn't very efficient at the moment.

FWIW, @jantimon's use case, as well as a similar use case I'm working on regarding SQL queries, do not make use of conditional types or literal type parsing. If I understand him correctly, the use of literal types we have in mind boils down to generating overloads, like so:

```ts
function template(strings: ["select ", " + 1"], parameter: number): number;
function template(strings: ["select ", " || ' world!'"], parameter: string): string;
...
```

No literal parsing required. I'd argue that tools like this will be the main use cases in the near future, since users will avoid libraries that do complex parsing within TypeScript because they are too slow.

**Comment by porsager** [comment]
What about those of us who just wants a way to map types against a tagged template literal function starting with specific words?

**Comment by ariofrio** [comment]
> What about those of us who just wants a way to map types against a tagged template literal function starting with specific words?

It sounds like that would be supported by this PR. And potentially simple enough that it wouldn't run into TypeScript performance constraints, and so, under my theory, be adopted widely by users and be one of "the main use cases", unlike more complex use cases that may need to wait for a more performant TypeScript to gain adoption.

My point is that performance is an important part of DX, and this is something that library authors and users already have to take into account. For example, [`uom`](https://github.com/iliekturtles/uom) for Rust is an awesome library, but it is not as widely used as it could be because it has quite slow compile times.

**Comment by sosoba** [comment]
> "npm:@typescript-deploys/pr-build@4.8.0-pr-49552-27"

Is it possible to release this pr with 5.7?

**Comment by porsager** [comment]
Pleeeeease reconsider merging this! I mean... Doom was implemented in Typescript but we can't do this? 😥

**Comment by sosoba** [comment]
Entry for posterity. Ts 5.9 still doesn't provide types for TemplateStringsArray :(

**Comment by diegoimbert** [comment]
Would really love it, it would enable me to have typed sql`QUERIES` in Windmill :)
